loginserver.js

```diff
@@ -17,20 +17,45 @@ const url = require('url');
 
 const FS = require('./fs');
 
-let TimeoutError = function (message) {
-	Error.captureStackTrace(this, TimeoutError);
-	this.name = "TimeoutError";
-	this.message = message || "";
-};
-TimeoutError.prototype = Object.create(Error.prototype);
-TimeoutError.prototype.constructor = TimeoutError;
-TimeoutError.prototype.toString = function () {
-	if (!this.message) return this.name;
-	return this.name + ": " + this.message;
-};
+/**
+ * A custom error thrown when a problem specific to the login server occurs,
+ * rather than the number of generic errors Node uses. Error's internal
+ * behaviour makes subclassing it require special treatment so it can act like
+ * an Error properly.
+ */
+class TimeoutError extends Error {
+	/**
+	 * Assigning this in the constructor would make the error print extra
+	 * properties unrelated to its stack.
+	 *
+	 * @return {string}
+	 */
+	get name() {
+		return 'TimeoutError';
+	}
+
+	/**
+	 * Stops instances from returning 'Error' when stringified.
+	 *
+	 * @return {string}
+	 */
+	get [Symbol.toStringTag]() {
+		return this.name;
+	}
+
+	/**
+	 * Stops instances from not being considered to be their actual type.
+	 *
+	 * @return {{new(message: [string]): TimeoutError}}
+	 */
+	static get [Symbol.species]() {
```

Comment:
It already does that now.

This actually ended up being slightly less performant:
```
instanceof TimeoutError x 26,871,105 ops/sec Â±0.93% (89 runs sampled)
instanceof TimeoutError2 x 23,736,389 ops/sec Â±0.40% (94 runs sampled)
Fastest is instanceof TimeoutError
```

And I was way, way off about `toString`, so I'll be changing that back to a regular method:
```
TimeoutError#toString x 59,094,784 ops/sec Â±0.67% (93 runs sampled)
TimeoutError2#[Symbol.toStringTag] x 2,683,727 ops/sec Â±1.24% (90 runs sampled)
Fastest is instanceof TimeoutError
```

After changing it back, it is faster:
```
TimeoutError#toString x 58,543,116 ops/sec Â±0.74% (92 runs sampled)
TimeoutError2#toString x 67,352,252 ops/sec Â±0.69% (93 runs sampled)
Fastest is instanceof TimeoutError2
```